Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Fin.VecNotation
+import Mathlib.Data.Rat.Order -- will be trimmed after deleting examples
+import Mathlib.Tactic.Linarith
+import Mathlib.Tactic.Cases
+import Mathlib.Data.Bool.Basic
+
+/-!
+
+# General-Valued Constraint Satisfaction Problems
+
+General-Valued CSP is a very broad class of problems in discrete optimization.
+General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finite-Valued CSP.
+
+## Main definitions
+* `ValuedCspTemplate`: A VCSP template; fixes a domain, a codomain, and allowed cost functions.
+* `ValuedCspTerm`: One summand in a VCSP instance; calls a concrete function from given template.
+* `ValuedCspTerm.evalSolution`: An evaluation of the VCSP term for given solution.
+* `ValuedCspInstance`: An instance of a VCSP problem over given template.
+* `ValuedCspInstance.evalSolution`: An evaluation of the VCSP instance for given solution.
+* `ValuedCspInstance.optimumSolution`: Is given solution a minimum of the VCSP instance?
+
+## References
+* [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
+   *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
+
+-/
+
+def n1ary_of_unary {α β : Type _} (f : α → β) : (Fin 1 → α) → β :=
+  fun a => f (a 0)
+
+def n2ary_of_binary {α β : Type _} (f : α → α → β) : (Fin 2 → α) → β :=
+  fun a => f (a 0) (a 1)
+
+/-- A template for a valued CSP problem with costs in `C`. -/
+structure ValuedCspTemplate (C : Type _) [LinearOrderedAddCommMonoid C] where
+  /-- Domain of "labels" -/
+  D : Type
```

Comment:
Since this is usually a finite set, i think it's fine to say this must live in the bottom type universe. Can you add a comment in the docstring (so that it's clear that it's intentional).